### PR TITLE
feat: Add Commands enum to decode prost messages to strong type

### LINKED ISSUE #3874: Make it easier to match FlightSQL messages
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
When implementing FlightSQL the do_get and get_flight_info gRPC methods are called for several of the message types. Thus those endpoints need to decode the incoming message and then do the appropriate action

However, the current way the messages are decoded from `prost::Any` means the only way I have figured out how to check dynamically for each message type is an `if` chain like this

```rust
        } else if let Some(decoded_cmd) = Any::unpack::<CommandGetCatalogs>(&msg)? {
          // act
        } else if let Some(decoded_cmd) = Any::unpack::<CommandGetDbSchemas>(&msg)? {
          // act
        } else if let Some(decoded_cmd) = Any::unpack::<ActionCreatePreparedStatementRequest>(&msg)?
        {
          // ...
        }
```

It would be great to have an enum so a rust `match` statement could be used

**Describe alternatives you've considered**
@stuartcarnie proposed the following in  https://github.com/influxdata/influxdb_iox/pull/7213/files#r1137963627:

 I created [a gist](https://gist.github.com/stuartcarnie/a121def085aaaf3f75ab5e592e4c4343#file-mod-rs-L230-L233) that extends [the sql crate](https://github.com/apache/arrow-rs/blob/9728c676b50b19c06643a23daba4aa4a1dc48055/arrow-flight/src/sql/mod.rs) in arrow-flight, that demonstrates an approach allowing us to remove all the `if Some(..)` to a single match:

```rust
match Commands::unpack(&msg)? {
    Commands::CommandStatementQuery(cmd) => Ok(Self::CommandStatementQuery(cmd)),
    Commands::CommandPreparedStatementQuery(CommandPreparedStatementQuery { prepared_statement_handle }) => Ok(Self::CommandPreparedStatementQuery(preparedstatementhandle::try_decode(prepared_statement_handle)?)),
    // etc
}
```

I extended the `prost_message_ext` macro to create a `Commands` enum with all the messages passed to the macro. The `Commands` macro has a static function, `unpack`, that is capable of unpacking any of the registered messages.

**Additional context**

### PR DESCRIPTION
**Pull request by stuartcarnie**
# Which issue does this PR close?

Closes #3874

# Rationale for this change
 
Per the issue, this PR introduces a `Commands` enum to allow client to match on all supported messages.

The following minor improvements were made to the `prost_message_ext` macro:

## 1. Generate a constant for the type URLs:

```rust
const ACTION_CLOSE_PREPARED_STATEMENT_REQUEST_TYPE_URL: &'static str = concat!( "type.googleapis.com/arrow.flight.protocol.sql.", stringify!( ActionClosePreparedStatementRequest ) );
const ACTION_CREATE_PREPARED_STATEMENT_REQUEST_TYPE_URL: &'static str = concat!( "type.googleapis.com/arrow.flight.protocol.sql.", stringify!( ActionCreatePreparedStatementRequest ) );
// Remainder omitted for brevity
```

## 2. `Commands` enum to unpack any of the supported FlightSQL commands:
 
```rust
pub enum Commands {
    ActionClosePreparedStatementRequest(ActionClosePreparedStatementRequest),
    ActionCreatePreparedStatementRequest(ActionCreatePreparedStatementRequest),
    // Remainder omitted for brevity
}
```

with an associated `unpack` function:

```rust
impl Commands {
    pub fn unpack(any: Any) -> Result<Commands, ArrowError> {
        match any.type_url.as_str() {
            ACTION_CLOSE_PREPARED_STATEMENT_REQUEST_TYPE_URL => {
                let m: ActionClosePreparedStatementRequest = Message::decode(&*any.value).map_err(|err| {
                    ArrowError::ParseError({
                        let res = ::alloc::fmt::format(format_args!("Unable to decode Any value: {err}", err = err));
                        res
                    })
                })?;
                Ok(Self::ActionClosePreparedStatementRequest(m))
            }
            ACTION_CREATE_PREPARED_STATEMENT_REQUEST_TYPE_URL => {
                let m: ActionCreatePreparedStatementRequest = Message::decode(&*any.value).map_err(|err| {
                    ArrowError::ParseError({
                        let res = ::alloc::fmt::format(format_args!("Unable to decode Any value: {err}", err = err));
                        res
                    })
                })?;
                Ok(Self::ActionCreatePreparedStatementRequest(m))
            }

            // Remainder omitted for brevity

            _ => Err(ArrowError::ParseError({
                let res = ::alloc::fmt::format(IntellijRustDollarCrate::__export::format_args!("Unable to decode Any value: {}", any.type_url));
                res
            }))
        }
    }
}
```

## 3. Update other generated code to use the shared constant for the type URL:

```rust
impl ProstMessageExt for ActionClosePreparedStatementRequest {
    fn type_url() -> &'static str {
        ACTION_CLOSE_PREPARED_STATEMENT_REQUEST_TYPE_URL
    }

    fn as_any(&self) -> Any {
        Any {
            type_url: <ActionClosePreparedStatementRequest>::type_url().to_string(),
            value: self.encode_to_vec().into(),
        }
    }
}
impl ProstMessageExt for ActionCreatePreparedStatementRequest {
    fn type_url() -> &'static str {
        ACTION_CREATE_PREPARED_STATEMENT_REQUEST_TYPE_URL
    }

    fn as_any(&self) -> Any {
        Any {
            type_url: <ActionCreatePreparedStatementRequest>::type_url().to_string(),
            value: self.encode_to_vec().into(),
        }
    }
}
// Remainder omitted for brevity
```

# What changes are included in this PR?

# Are there any user-facing changes?

API improvements for users of the FlightSQL module.

### CODE REVIEW

**Comment by alamb** [comment]
I plan to review this shortly

**Review by alamb — COMMENTED** [review]
Thank you @stuartcarnie  -- this is very helpful ❤️  -- I really like it

Some things I noticed while playing with this:

1. In order to make this code more "discoverable"  I think it would be great to add some additional documentation / examples and update the examples. I am working on a PR to do so
2. . It would be nice if the FlightSQL implementation used `FlightError` rather than ArrowError (not caused by this PR). I will file a follow on PR to track this

File: arrow-flight/src/sql/mod.rs [hunk 1]
```diff
@@ -71,22 +72,60 @@ pub trait ProstMessageExt: prost::Message + Default {
...
-                fn type_url() -> &'static str {
-                    concat!("type.googleapis.com/arrow.flight.protocol.sql.", stringify!($name))
+    ($($name:tt,)*) => {
+        paste! {
+            $(
+            const [<$name:snake:upper _TYPE_URL>]: &'static str = concat!("type.googleapis.com/arrow.flight.protocol.sql.", stringify!($name));
+            )*
+
+            as_item! {
+                pub enum Commands {
```
Comment (alamb) [inline arrow-flight/src/sql/mod.rs:93]: Screenshot 2023-03-20 at 4 40 12 PM

File: arrow-flight/src/sql/mod.rs [hunk 2]
```diff
@@ -71,22 +72,60 @@ pub trait ProstMessageExt: prost::Message + Default {
...
+    };
+}
+
 macro_rules! prost_message_ext {
-    ($($name:ty,)*) => {
-        $(
-            impl ProstMessageExt for $name {
-                fn type_url() -> &'static str {
-                    concat!("type.googleapis.com/arrow.flight.protocol.sql.", stringify!($name))
+    ($($name:tt,)*) => {
```
Comment (alamb) [inline arrow-flight/src/sql/mod.rs:86]: What do you think about calling this `Command` to be more idiomatic (enums normally aren't plural) or `FlightSqlCommand` (if we want to be more explicit)

**Comment by alamb** [comment]
Here are some suggested improvements: https://github.com/stuartcarnie/arrow-rs/pull/2# (docs + revamp some of the server.rs code to use the new dispatch logic 👍 )

File: arrow-flight/src/sql/mod.rs (hunk 1 above)
Comment (stuartcarnie) [reply arrow-flight/src/sql/mod.rs:93]: Good call – do you think adding some docs to the enum and then describing using the `TryFrom<Any>` trait to decoding them?

File: arrow-flight/src/sql/mod.rs (hunk 2 above)
Comment (stuartcarnie) [reply arrow-flight/src/sql/mod.rs:86]: Done – I've renamed it to `Command`

File: arrow-flight/src/sql/mod.rs [hunk 3]
```diff
@@ -71,22 +72,60 @@ pub trait ProstMessageExt: prost::Message + Default {
...
+                    $($name($name),)*
                 }
+            }
 
-                fn as_any(&self) -> Any {
-                    Any {
-                        type_url: <$name>::type_url().to_string(),
-                        value: self.encode_to_vec().into(),
+            impl Commands {
+                pub fn unpack(any: Any) -> Result<Commands, ArrowError> {
```
[the remark this answers is not in the document]
Comment (stuartcarnie) [reply arrow-flight/src/sql/mod.rs:120]: Agreed – I've removed unpack completely, as `Command::try_from` or `any.try_into()` is much better 👍🏻

File: arrow-flight/src/sql/mod.rs [hunk 4]
```diff
@@ -71,22 +72,104 @@ pub trait ProstMessageExt: prost::Message + Default {
...
+                        Self::$name(cmd) => cmd.as_any(),
+                        )*
+                    }
                 }
 
-                fn as_any(&self) -> Any {
-                    Any {
-                        type_url: <$name>::type_url().to_string(),
-                        value: self.encode_to_vec().into(),
+                /// Get the URL for the command.
```
Comment (alamb) [inline arrow-flight/src/sql/mod.rs:129]: 👍  this is great

**Review by alamb — APPROVED** [review]
Thanks @stuartcarnie -- I think this is looking great

One thing I would really like to do is to translate 

```rust
    async fn do_get(
```

in `arrow-flight/src/sql/server.rs` to use this new pattern too (mostly as an exercise in trying out the API). I ran out of time last night.

**Comment by alamb** [comment]
I pushed a few minor doc updates directly to this branch.

File: arrow-flight/src/sql/server.rs [hunk 5]
```diff
@@ -315,90 +315,57 @@ where
...
+            | Command::TicketStatementQuery(_)
+            | Command::ActionCreatePreparedStatementResult(_) => {
+                Err(Status::unimplemented(format!(
+                    "get_flight_info: The defined request is invalid: {}",
+                    command.type_url()
+                )))
+            }
+            Command::Unknown(Any { ref type_url, .. }) => Err(Status::invalid_argument(
+                format!("get_flight_info: The defined request is invalid: {type_url}"),
+            )),
```
Comment (stuartcarnie) [inline arrow-flight/src/sql/server.rs:367]: @alamb I wasn't sure if this was deliberately an `unimplemented` error, so I did not combine the `Unknown` variant, which returns an `invalid_argument` status.

**Comment by stuartcarnie** [comment]
@alamb see ea18fa9697d1face7945e40dc6542b59c3890614 for the addition of the `Unknown(Any)` variant, which was a great suggestion. I've also updated `do_get` and `do_put`, and the simple test now passes ✅

File: arrow-flight/src/sql/server.rs [hunk 6]
```diff
@@ -315,90 +315,46 @@ where
...
+            Command::CommandGetImportedKeys(token) => {
+                self.get_flight_info_imported_keys(token, request).await
+            }
+            Command::CommandGetCrossReference(token) => {
+                self.get_flight_info_cross_reference(token, request).await
+            }
+            cmd => Err(Status::unimplemented(format!(
+                "get_flight_info: The defined request is invalid: {}",
+                cmd.type_url()
+            ))),
```
Comment (stuartcarnie) [inline arrow-flight/src/sql/server.rs:356]: @alamb I combined all remaining variants into a single error. I left it as `unimplemented`, as it is unclear to me if there are additional commands that may be implemented in the future. If not, perhaps `invalid_argument` is a more appropriate response, however, given that is a change that could break downstream clients, I've left it.

File: arrow-flight/src/sql/server.rs (hunk 5 above)
Comment (stuartcarnie) [reply arrow-flight/src/sql/server.rs]: Disregard, as I've combined remaining cases.

**Comment by alamb** [comment]
Thanks -- I plan to come back to this PR over the weekend but got overloaded with other things this week

**Comment by alamb** [comment]
I am sorry for the delay -- I will get this done / merged this week

File: arrow-flight/src/sql/mod.rs [hunk 7]
```diff
@@ -190,4 +279,27 @@ mod tests {
...
+        };
+        let any = Any::pack(&query).unwrap();
+        let cmd: Command = any.try_into().unwrap();
+
+        assert!(matches!(cmd, Command::CommandStatementQuery(_)));
+        assert_eq!(cmd.type_url(), COMMAND_STATEMENT_QUERY_TYPE_URL);
+
+        // Unknown variant
+
+        let any = Any {
```
Comment (alamb) [inline arrow-flight/src/sql/mod.rs:296]: 👍

**Review by alamb — APPROVED** [review]
Thanks @stuartcarnie  -- I reviewed this again and I think it looks great. I merged up from main to resolve a conflict and plan to merge this PR in when it passes CI

**Comment by alamb** [comment]
Github seems to have some issue -- restarting

Screenshot 2023-04-03 at 3 05 12 PM

**Comment by alamb** [comment]
I propose some additional documentation for this feature here https://github.com/apache/arrow-rs/pull/4012

**Pull request merged**